### PR TITLE
refactor: available balance

### LINKED ISSUE #3569: Improve balances UX
With this enhancement https://github.com/hirosystems/wallet/pull/3525

Users can see the error in the image below (insufficient balance...):
When they have 350 STX in their wallet, and a pending transaction that will use 150 STX (100 STX transfer and 50 STX fee). There is only 200 STX available. 

Showing two different values will lead to questions. Lets answer the likely question (A) or prevent the questions (B)?
A. By explaining the difference between available balance (shown on error) and the balance shown in the lower right corner.
I.e "Available balance takes into account pending transfers of STX and pending fees"
B. Or by showing available balance instead?
(...and other transactions with postconditions that suggest outgoing STX - https://github.com/hirosystems/wallet/issues/3568).

image

Because of a pending transaction like this (100 STX going out, and 50 STX fee)
image

### PR DESCRIPTION
**Pull request by alter-eggo**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/wallet/actions/runs/4870759185).

Refactor user available balance and subtract microblock txs from it

### CODE REVIEW

File: src/app/query/stacks/microblock/microblock.query.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { TokenTransferTransaction } from '@stacks/stacks-blockchain-api-types';
+import { useQuery } from '@tanstack/react-query';
+
+import { useStacksClientUnanchored } from '@app/store/common/api-clients.hooks';
+
+export function useCurrentAccountMicroblockQuery(address: string) {
+  const client = useStacksClientUnanchored();
+
+  async function accountMicroblockFetcher() {
+    const txs = await client.microblocksApi.getUnanchoredTxs({});
```
Comment (alter-eggo) [inline src/app/query/stacks/microblock/microblock.query.ts:13]: not sure if I've chosen right api method here

File: src/app/common/hooks/balance/use-stx-available-balance.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { useMemo } from 'react';
+
+import { createMoney } from '@shared/models/money.model';
+
+import { subtractMoney } from '@app/common/money/format-money';
+import { useCurrentStacksAccountAnchoredBalances } from '@app/query/stacks/balance/stx-balance.hooks';
+import { useCurrentAccountMempoolTransactionsBalance } from '@app/query/stacks/mempool/mempool.hooks';
+import { useCurrentAccountMicroblockTransactionsBalance } from '@app/query/stacks/microblock/microblock.hooks';
+
+export function useStxAvailableBalance() {
```
Comment (fbwoolf) [inline src/app/common/hooks/balance/use-stx-available-balance.ts]: It is getting confusing to me in the code about what `availableStx` balance means? If you are going to reuse, and now subtract out pending/microblock txs, should we rename the previous `availableStx` on the balance to something like `unlockedStx`? Just a thought bc I had to lookup and remind myself what was actually different about `balances?.stx.availableStx`. Or, we could rename this hook to something descriptive like, `useStxAvailableBalanceMinusPendingAndMicroblockTxs`? 😆

File: src/app/query/stacks/microblock/microblock.query.ts (hunk 1 above)
Comment (fbwoolf) [reply src/app/query/stacks/microblock/microblock.query.ts:13]: We haven't used this endpoint before, but it does appear to return microblock txs so if is working seems ok? I think before we were filtering other tx queries by `transaction.tx_status === 'success' && transaction.is_unanchored` which would yield a mircoblock transaction.

Comment (kyranjamie) [reply src/app/query/stacks/microblock/microblock.query.ts:13]: This fetched all of them, which could be loads (isn't this a paginated API resp?).

Don't we get the same info from the balance response? We have both 

`useUnanchoredStacksAccountBalanceQuery`
`useAnchoredStacksAccountBalanceQuery`

Afaik, the former should get the same result 

(sorry for slow review)

File: src/app/common/hooks/balance/use-stx-available-balance.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  const availableStxBalance = balances?.stx.availableStx ?? createMoney(0, 'STX');
+
+  const pendingTxsBalance = useCurrentAccountMempoolTransactionsBalance();
+  const microblockTxsBalance = useCurrentAccountMicroblockTransactionsBalance();
+
+  return useMemo(
+    () =>
+      subtractMoney(subtractMoney(availableStxBalance, pendingTxsBalance), microblockTxsBalance),
+    [availableStxBalance, pendingTxsBalance, microblockTxsBalance]
```
Comment (kyranjamie) [inline src/app/common/hooks/balance/use-stx-available-balance.ts]: same point here on using `select` with react query might be better

File: src/app/common/hooks/balance/use-stx-available-balance.ts (hunk 2 above)
Comment (alter-eggo) [reply src/app/common/hooks/balance/use-stx-available-balance.ts]: not sure at all, need some help with naming here @kyranjamie @markmhx

Comment (markmhendrickson) [reply src/app/common/hooks/balance/use-stx-available-balance.ts]: Agreed naming is getting confused here. Let's sort out in this doc, which also touches on a possible implementation change for `.availableStx` here specifically https://www.notion.so/Balances-UX-905c93b633b84e05a1428c1a9087ab82

File: src/app/common/money/format-money.ts [hunk 4]
```diff
@@ -32,3 +32,8 @@ export function subtractMoney(xAmount: Money, yAmount: Money) {
   if (xAmount.symbol !== yAmount.symbol) throw new Error('Cannot subtract different currencies');
   return createMoney(xAmount.amount.minus(yAmount.amount), xAmount.symbol, xAmount.decimals);
 }
+
+export function additionMoney(xAmount: Money, yAmount: Money) {
```
Comment (kyranjamie) [inline src/app/common/money/format-money.ts]: Would keep it a verb, and go with `addMoney`

File: src/app/query/stacks/microblock/microblock.hooks.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+export function useCurrentAccountMicroblockTransactionsBalance() {
+  const stxAddress = useCurrentAccountStxAddressState();
+  const { data: microblockTxs } = useCurrentAccountMicroblockQuery(stxAddress);
+
+  const senderMicroblockTxs = microblockTxs?.filter(tx => tx.sender_address === stxAddress) || [];
+
+  const senderMicroblockTxsBalance = senderMicroblockTxs.reduce(
+    (acc, tx) => acc.plus(tx.token_transfer.amount),
+    new BigNumber(0)
+  );
```
Comment (kyranjamie) [inline src/app/query/stacks/microblock/microblock.hooks.ts]: see `sumNumbers` helper, maybe we could make a wrapper that takes `Money`?

File: src/app/query/stacks/microblock/microblock.query.ts (hunk 1 above)
Comment (alter-eggo) [reply src/app/query/stacks/microblock/microblock.query.ts:13]: seems it is not paginated? https://docs.hiro.so/api?_gl=1*1reabcz*_ga*MTA1Mjk0OTM1OC4xNjczMzczOTQ4*_ga_NB2VBT0KY2*MTY4Mjk0MDI2Ni42My4wLjE2ODI5NDAyNjYuMC4wLjA.#tag/Microblocks/operation/get_unanchored_txs

Comment (kyranjamie) [reply src/app/query/stacks/microblock/microblock.query.ts:13]: Great, let's stick with this approach then!

**Comment by markmhendrickson** [comment]
"Available balance" for send flow with new tooltip

File: src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx [hunk 6]
```diff
@@ -28,7 +31,20 @@ export function FormFooter(props: { balance: Money }) {
       <Box mt="loose" px="loose">
         <PreviewButton />
         <SpaceBetween>
-          <Caption>Balance</Caption>
+          <Flex alignItems="center">
+            <Caption mr="tight">Available balance</Caption>
+            <Tooltip placement="top" label={balanceTooltipLabel}>
+              <Stack>
```
Comment (fbwoolf) [inline src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx:37]: Is the `Stack` needed here?

File: src/app/query/stacks/balance/stx-balance.hooks.ts [hunk 7]
```diff
@@ -23,10 +23,12 @@ export function parseBalanceResponse(balances: AddressBalanceResponse) {
     ])
   ) as Record<AccountBalanceStxKeys, Money>;
 
-  const stx: AccountStxBalanceBigNumber & { availableStx: Money } = {
+  const stx: AccountStxBalanceBigNumber & { availableStx: Money; unlockedStx: Money } = {
     ...balances.stx,
     ...stxMoney,
+    balance: createMoney(stxMoney.balance.amount, 'STX'),
     availableStx: createMoney(stxMoney.balance.amount.minus(stxMoney.locked.amount), 'STX'),
```
Comment (fbwoolf) [inline src/app/query/stacks/balance/stx-balance.hooks.ts]: Shouldn't L30 be deleted now in favor of using `unlockedStx` throughout the code?

**Review by fbwoolf — APPROVED** [review]
Looks great @alter-eggo, thanks for addressing the naming here!

File: src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx (hunk 6 above)
Comment (alter-eggo) [reply src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx:37]: yeah, without it tooltip is not showing, it's used so in many places, e.g. in `fee-estimate-select.layout.tsx`

File: src/app/query/stacks/balance/stx-balance.hooks.ts (hunk 7 above)
Comment (alter-eggo) [reply src/app/query/stacks/balance/stx-balance.hooks.ts]: yes, fixed.
my first intention was to do it in separate pr, but now think its better to do it here

File: src/app/common/hooks/balance/stx/use-stx-available-balance.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  const unlockedStxBalance = balances?.stx.unlockedStx;
+  const stxOutboundValue = useStxOutboundValue();
+
+  return useMemo(
+    () =>
+      isDefined(unlockedStxBalance)
+        ? subtractMoney(unlockedStxBalance, stxOutboundValue)
+        : createMoney(0, 'STX'),
+    [unlockedStxBalance, stxOutboundValue]
+  );
```
Comment (kyranjamie) [inline src/app/common/hooks/balance/stx/use-stx-available-balance.ts]: How does loading work? Does it flash from 0 to balance? Wonder if we should also use select here? What do you think?

File: src/app/common/hooks/balance/stx/use-stx-balance.ts [hunk 9]
```diff
@@ -5,6 +5,7 @@ import { i18nFormatCurrency } from '@app/common/money/format-money';
 import { useCryptoCurrencyMarketData } from '@app/query/common/market-data/market-data.hooks';
 import { useStacksAnchoredCryptoCurrencyAssetBalance } from '@app/query/stacks/balance/stacks-ft-balances.hooks';
 
+// ts-unused-exports:disable-next-line
```
Comment (kyranjamie) [inline src/app/common/hooks/balance/stx/use-stx-balance.ts]: if it isn't used, can we kill it entirely?

File: src/app/query/stacks/balance/stx-balance.hooks.ts [hunk 10]
```diff
@@ -23,10 +23,11 @@ export function parseBalanceResponse(balances: AddressBalanceResponse) {
     ])
   ) as Record<AccountBalanceStxKeys, Money>;
 
-  const stx: AccountStxBalanceBigNumber & { availableStx: Money } = {
+  const stx: AccountStxBalanceBigNumber & { unlockedStx: Money } = {
     ...balances.stx,
     ...stxMoney,
-    availableStx: createMoney(stxMoney.balance.amount.minus(stxMoney.locked.amount), 'STX'),
+    balance: createMoney(stxMoney.balance.amount, 'STX'),
+    unlockedStx: createMoney(stxMoney.balance.amount.minus(stxMoney.locked.amount), 'STX'),
```
Comment (kyranjamie) [inline src/app/query/stacks/balance/stx-balance.hooks.ts:30]: These are already `Money`, can't se use the new helper?

File: src/app/components/balance-stx.tsx [hunk 11]
```diff
@@ -14,7 +14,7 @@ export function StxBalance(props: BalanceProps) {
   const balance = useMemo(
     () =>
       stacksValue({
-        value: balances?.stx?.availableStx.amount ?? 0,
+        value: balances?.stx?.unlockedStx.amount ?? 0,
         withTicker: true,
```
Comment (kyranjamie) [inline src/app/components/balance-stx.tsx:18]: For another time, but would like to remove `stacksValue` in favour of formatters that work with `Money`

**Review by kyranjamie — APPROVED** [review]
Great work @alter-eggo 

Main thoughts on how this could be improved surrounds loading of balances. We don't have a loading indicator for balances, but if we wanted to add one, how would we access the `isLoading` to show an indicator in the UI?

Defaulting to `0` for the unloaded state is helpful and often good enough, but it isn't always the behaviour we want. How might we structure the hooks in such that we can default to 0 if needed, but we don't always make that assumption?

File: src/app/query/stacks/balance/stx-balance.hooks.ts (hunk 10 above)
Comment (alter-eggo) [reply src/app/query/stacks/balance/stx-balance.hooks.ts:30]: I used helper, as there is no `decimals` in balance field

File: src/app/common/hooks/balance/stx/use-stx-available-balance.ts (hunk 8 above)
Comment (alter-eggo) [reply src/app/common/hooks/balance/stx/use-stx-available-balance.ts]: I restructured all this hooks to return loading state

**Review by markmhendrickson — APPROVED** [review]
The STX balance seems to react as expected when I have outbound pending transactions.

Note however that the account's top-level fiat USD balance doesn't seem to change and reflect the pending balance change. Rather it only updates when the STX transaction confirms.

Also it seems that we're not updating the BTC balance to reflect outbound pending amounts in parallel with STX ones. 

Should we merge this in and tackle these two items separately, or easy enough to include in this PR? I presume best to merge and tackle separately.

**Comment by alter-eggo** [comment]
I'd prefer do it separately

**Pull request merged**